File: mobile/src/main/java/com/ataulm/chunks/AndroidChunksPresenter.java
package com.ataulm.chunks;

import com.ataulm.AndroidLog;
import com.ataulm.Event;
import com.ataulm.Log;
import com.ataulm.LoggingObserver;
import com.ataulm.Optional;

import rx.Subscription;

class AndroidChunksPresenter implements ChunksPresenter {

    private final ChunksService chunksService;
    private final ChunksView chunksView;

    private Subscription subscription;

    AndroidChunksPresenter(ChunksService chunksService, ChunksView chunksView) {
        this.chunksService = chunksService;
        this.chunksView = chunksView;
    }

    @Override
    public void startPresenting() {
        safeUnsubscribeFrom(subscription);
        subscription = chunksService.fetchEntries()
                .subscribe(
                        new ChunksObserver(
                                new AndroidLog(),
                                new AndroidItemUserInteractions(chunksService),
                                new AndroidItemInputUserInteractions(chunksService)
                        )
                );
    }

    @Override
    public void onExternalShareText(String text) {
        chunksService.createEntry(Item.createNew(text.trim()), Day.TODAY);
    }

    @Override
    public void stopPresenting() {
        safeUnsubscribeFrom(subscription);
        chunksService.persist();
    }

    private static void safeUnsubscribeFrom(Subscription subscription) {
        if (subscription != null && !subscription.isUnsubscribed()) {
            subscription.unsubscribe();
        }
    }

    private class ChunksObserver extends LoggingObserver<Event<Chunks>> {

        private final ItemUserInteractions entryViewUserInteractions;
        private final ItemInputUserInteractions itemInputUserInteractions;

        public ChunksObserver(
                Log log,
                ItemUserInteractions entryViewUserInteractions,
                ItemInputUserInteractions itemInputUserInteractions
        ) {
            super(log);
            this.entryViewUserInteractions = entryViewUserInteractions;
            this.itemInputUserInteractions = itemInputUserInteractions;
        }

        @Override
        public void onNext(Event<Chunks> event) {
            Optional<Chunks> data = event.getData();
            if (data.isPresent()) {
                Chunks chunks = data.get();
                chunksView.display(chunks, entryViewUserInteractions, itemInputUserInteractions);
            } else {
                // TODO other cases like loading (empty, not empty), empty state, error (empty, not empty)
            }
        }

    }

}

File: mobile/src/main/java/com/ataulm/chunks/AndroidChunksPresenter.kt
package com.ataulm.chunks

import com.ataulm.AndroidLog
import com.ataulm.Event
import com.ataulm.Log
import com.ataulm.LoggingObserver

import rx.Subscription

internal class AndroidChunksPresenter(private val chunksService: ChunksService, private val chunksView: ChunksView) : ChunksPresenter {

    private var subscription: Subscription? = null

    override fun startPresenting() {
        safeUnsubscribeFrom(subscription)
        subscription = chunksService.fetchEntries()
                .subscribe(
                        ChunksObserver(
                                AndroidLog(),
                                AndroidItemUserInteractions(chunksService),
                                AndroidItemInputUserInteractions(chunksService)
                        )
                )
    }

    override fun onExternalShareText(text: String) {
        chunksService.createEntry(Item.createNew(text.trim { it <= ' ' }), Day.TODAY)
    }

    override fun stopPresenting() {
        safeUnsubscribeFrom(subscription)
        chunksService.persist()
    }

    private fun safeUnsubscribeFrom(subscription: Subscription?) {
        if (subscription != null && !subscription.isUnsubscribed) {
            subscription.unsubscribe()
        }
    }

    private inner class ChunksObserver internal constructor(
            log: Log,
            private val entryViewUserInteractions: ItemUserInteractions,
            private val itemInputUserInteractions: ItemInputUserInteractions
    ) : LoggingObserver<Event<Chunks>>(log) {

        override fun onNext(event: Event<Chunks>) {
            val data = event.data
            if (data != null) {
                display(data)
            } else {
                // TODO other cases like loading (empty, not empty), empty state, error (empty, not empty)
                display(Chunks.empty(ChunkDate.create(SystemClock())))
            }
        }

        private fun display(chunks: Chunks) {
            chunksView.display(chunks, entryViewUserInteractions, itemInputUserInteractions)
        }
    }
}
